### PR TITLE
Garbage Collection for `AutoGradTensor`

### PR DESCRIPTION
**Pull request by Syzygianinfern0**
Issue addressed: https://www.diffchecker.com/7jVaQe96

Code for testing
```python
import torch as th
import torch.optim as optim
import pprint
import syft as sy

hook = sy.TorchHook(th) #, verbose=True)
bob = sy.VirtualWorker(hook, id="bob") #, verbose=True)
alice = sy.VirtualWorker(hook, id="alice") #, verbose=True)
crypto_provider = sy.VirtualWorker(hook, id="james") #, verbose=True)

torch = th
syft = sy

a = torch.ones(1, 5)
b = torch.ones(1, 5)
a = a.encrypt(workers=[alice, bob], crypto_provider=crypto_provider, requires_grad=True)
b = b.encrypt(workers=[alice, bob], crypto_provider=crypto_provider, requires_grad=True)


class Classifier(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(Classifier, self).__init__()
        self.fc = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        logits = self.fc(x)
        return logits


classifier = Classifier(in_features=5, out_features=5)
model = classifier.fix_prec().share(alice, bob, crypto_provider=crypto_provider, requires_grad=True)
opt = optim.SGD(params=model.parameters(), lr=0.1).fix_precision()

for i in range(3):
    print("=" * 10 + f"Iter{i + 1}" + "=" * 10)
    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")
    preds = classifier(a)
    loss = ((b - preds) ** 2).sum()
    opt.zero_grad()
    loss.backward()
    opt.step()
    print(f"Loss : {loss.get().float_prec()}")
    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")

print("~" * 10 + "Done" + "~" * 10)

```

```
==========Iter1==========
Alice: 4
Bob: 4
Loss : 6.130000114440918
Alice: 11
Bob: 11
==========Iter2==========
Alice: 11
Bob: 11
Loss : 0.24400000274181366
Alice: 11
Bob: 11
==========Iter3==========
Alice: 11
Bob: 11
Loss : 0.009999999776482582
Alice: 11
Bob: 11
~~~~~~~~~~Done~~~~~~~~~~
```

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/gradients.py [hunk 1]
```diff
@@ -147,14 +147,14 @@ def gradient(self, grad):
...
-    def gradient(self, grad):
-        grad_self_ = grad / (2 * self.result)
-        return (grad_self_,)
+# class SqrtBackward(GradFunc):
+#     def __init__(self, self_):
+#         super().__init__(self, self_)
+#         self.self_ = self_
+#
+#     def gradient(self, grad):
+#         grad_self_ = grad / (2 * self.result) TODO: Broken as of Garbage Collection for `AutoGradTensor` (#3387)
```
Comment (karlhigley) [inline syft/frameworks/torch/tensors/interpreters/gradients.py:156]: Probably want to find a way to un-break this before merging.

File: syft/workers/base.py [hunk 2]
```diff
@@ -314,7 +314,11 @@ def recv_msg(self, bin_message: bin) -> bin:
         msg = sy.serde.deserialize(bin_message, worker=self)
 
         if self.verbose:
-            print(f"worker {self} received {type(msg).__name__} {msg.contents}")
+            print(
+                f"worker {self} received {type(msg).__name__} {msg.contents}"
+                if hasattr(msg, "contents")
+                else f"worker {self} received {type(msg).__name__}"
+            )
```
Comment (karlhigley) [inline syft/workers/base.py:321]: 👍

**Comment by AlanAboudib** [comment]
Your review will be appreciated @LaRiffle  :)
